src/ParallelAlgorithms/LinearSolver/Multigrid/Actions/RestrictFields.hpp

```diff
@@ -0,0 +1,221 @@
+ * \tparam OptionsGroup The option group identifying the multigrid solver
+ * \tparam FieldsAreMassiveTag A boolean tag in the DataBox that indicates
+ * whether or not the `FieldsTags` have already been multiplied by the mass
+ * matrix. This setting influences the way the fields are projected. In
+ * particular, the mass matrix already includes a Jacobian factor, so the
+ * difference in size between the parent and the child element is already
+ * accounted for.
+ * \tparam ReceiveTags The projected fields will be stored in these tags
+ * (default: `FieldsTags`).
+ */
+template <typename FieldsTags, typename OptionsGroup,
+          typename FieldsAreMassiveTag, typename ReceiveTags = FieldsTags>
+struct SendFieldsToCoarserGrid;
+
+/// \cond
+template <typename... FieldsTags, typename OptionsGroup,
+          typename FieldsAreMassiveTag, typename... ReceiveTags>
+struct SendFieldsToCoarserGrid<tmpl::list<FieldsTags...>, OptionsGroup,
+                               FieldsAreMassiveTag,
+                               tmpl::list<ReceiveTags...>> {
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            size_t Dim, typename ActionList, typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ElementId<Dim>& element_id, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    // Skip restriction on coarsest level
+    const auto& parent_id = db::get<Tags::ParentId<Dim>>(box);
+    if (not parent_id.has_value()) {
+      return {std::move(box)};
+    }
+
+    const size_t iteration_id =
+        db::get<Convergence::Tags::IterationId<OptionsGroup>>(box);
+    if (UNLIKELY(db::get<logging::Tags::Verbosity<OptionsGroup>>(box) >=
+                 ::Verbosity::Debug)) {
+      Parallel::printf("%s %s(%zu): Send fields to coarser grid\n", element_id,
+                       Options::name<OptionsGroup>(), iteration_id);
+    }
+
+    // Restrict the fields to the coarser (parent) grid.
+    // We restrict before sending the data so the restriction operation is
+    // parellelized. The parent only needs to sum up all child contributions.
+    const auto& mesh = db::get<domain::Tags::Mesh<Dim>>(box);
+    const auto& parent_mesh = db::get<Tags::ParentMesh<Dim>>(box);
+    ASSERT(
+        parent_mesh.has_value(),
+        "Should have a parent mesh, because a parent ID is set. This element: "
+            << element_id << ", parent element: " << *parent_id);
+    const auto child_size =
+        domain::child_size(element_id.segment_ids(), parent_id->segment_ids());
+    bool massive = false;
+    if constexpr (not std::is_same_v<FieldsAreMassiveTag, void>) {
+      massive = db::get<FieldsAreMassiveTag>(box);
+    }
+    tuples::TaggedTuple<ReceiveTags...> restricted_fields{};
```

Comment:
well the `ReceiveTags` be non-Tensor<DataVector> in the future? If not, you might want to consider using a `Variables` instead. I think you'd save not just on memory allocations but would also get better throughput on the matrix operations (`apply_matrices` does 2 memory allocations per call)

Reply:
The `ReceiveTags` will actually hold `Variables` :) The action supports restricting _multiple_ Variables. Combining the Variables into one could possibly improve performance, but this currently only happens ~5 times in a simulation (per element per multigrid-level), namely once per non-linear solver iteration.

I added a note to the docs on the requirements for the `ReceiveTags` (they must support `apply_matrices` and addition).

Reply:
Okay, sounds good. Thanks!